File: platform/android-customer/app/src/main/java/ca/shawcliffe/tomsproduce/PushManager.kt
package ca.shawcliffe.tomsproduce

import android.content.Context
import android.content.SharedPreferences
import com.google.firebase.messaging.FirebaseMessaging
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch
import kotlinx.coroutines.tasks.await

/**
 * Customers don't authenticate (see platform/ARCHITECTURE-MAP.md), so there's
 * no session to hang a push token off of. Instead we persist the customer_id
 * returned by POST /api/signup locally, and only register the device token
 * with the server once both pieces exist. Mirrors ios-customer/PushManager.swift,
 * but over FCM instead of direct APNs.
 *
 * KNOWN GAP: no Firebase project/google-services.json exists yet (see
 * README), so FirebaseMessaging.getInstance() throws until one is added —
 * every call here is best-effort and silently no-ops on failure, same as the
 * iOS app's "retries next launch" behaviour. The server's
 * POST /api/push/register and /api/broadcast/push also only handle
 * customers.apns_token today; sending this as an FCM token there is a no-op
 * until the server is updated to store/send via customers.fcm_token too.
 */
object PushManager {
    private const val PREFS_NAME = "shawcliffe.push"
    private const val CUSTOMER_ID_KEY = "customerId"

    private lateinit var prefs: SharedPreferences
    private var deviceToken: String? = null
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)

    fun init(context: Context) {
        prefs = context.applicationContext.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
    }

    var customerId: String?
        get() = prefs.getString(CUSTOMER_ID_KEY, null)
        set(value) = prefs.edit().putString(CUSTOMER_ID_KEY, value).apply()

    /** Call on every app launch — refreshes the token if Firebase is configured; no-ops otherwise. */
    suspend fun requestPermissionAndRegister() {
        try {
            val token = FirebaseMessaging.getInstance().token.await()
            didReceiveDeviceToken(token)
        } catch (e: Exception) {
            // Firebase not configured, or user hasn't granted POST_NOTIFICATIONS yet.
        }
    }

    fun didReceiveDeviceToken(token: String) {
        deviceToken = token
        registerIfPossible()
    }

    /** Call after a successful signup, since that's the only moment the app learns its own customer_id. */
    fun didCompleteSignup(customerId: String) {
        this.customerId = customerId
        registerIfPossible()
    }

    private fun registerIfPossible() {
        val token = deviceToken ?: return
        val id = customerId ?: return
        scope.launch {
            try {
                APIClient.post<PushRegisterRequest, PushRegisterResponse>(
                    "api/push/register",
                    PushRegisterRequest(Config.CLIENT_ID, id, token),
                )
            } catch (e: Exception) {
                // Best-effort — deviceToken/customerId persist, so this retries next launch.
            }
        }
    }
}

File: clients/little-dough-bread-co/ios-customer/ShawcliffeCustomer/PushManager.swift
import Foundation
import UIKit
import UserNotifications

/// Customers don't authenticate (see platform/ARCHITECTURE-MAP.md), so there's
/// no session to hang a push token off of. Instead we persist the customer_id
/// returned by POST /api/signup locally, and only register the device token
/// with the server once both pieces exist.
@MainActor
final class PushManager: ObservableObject {
    static let shared = PushManager()

    private let customerIdKey = "shawcliffe.customerId"
    private var deviceToken: String?

    private init() {}

    var customerId: String? {
        get { UserDefaults.standard.string(forKey: customerIdKey) }
        set { UserDefaults.standard.setValue(newValue, forKey: customerIdKey) }
    }

    /// Call on every app launch. Prompts only the first time; on subsequent
    /// launches this silently re-registers (refreshing the token) if the
    /// user already granted permission, and does nothing if they declined.
    func requestPermissionAndRegister() async {
        let center = UNUserNotificationCenter.current()
        let settings = await center.notificationSettings()

        switch settings.authorizationStatus {
        case .authorized, .provisional:
            UIApplication.shared.registerForRemoteNotifications()
        case .notDetermined:
            do {
                let granted = try await center.requestAuthorization(options: [.alert, .sound, .badge])
                if granted {
                    UIApplication.shared.registerForRemoteNotifications()
                }
            } catch {
                // User can still enable later from Settings.
            }
        case .denied, .ephemeral:
            break
        @unknown default:
            break
        }
    }

    func didReceiveDeviceToken(_ tokenData: Data) {
        deviceToken = tokenData.map { String(format: "%02x", $0) }.joined()
        Task { await registerIfPossible() }
    }

    /// Call after a successful signup, since that's the only moment the app
    /// learns its own customer_id.
    func didCompleteSignup(customerId: String) {
        self.customerId = customerId
        Task { await registerIfPossible() }
    }

    private func registerIfPossible() async {
        guard let deviceToken, let customerId else { return }
        do {
            _ = try await APIClient.post(
                path: "api/push/register",
                body: PushRegisterRequest(client_id: Config.clientId, customer_id: customerId, token: deviceToken),
                as: PushRegisterResponse.self
            )
        } catch {
            // Best-effort — deviceToken/customerId persist, so this retries next launch.
        }
    }
}
